src/lem/circuit.rs

```diff
@@ -56,9 +64,117 @@ use super::{
     Block, Ctrl, Func, Op, Tag, Var,
 };
 
+#[derive(Clone)]
+pub struct MultiFrame<'a, F: LurkField> {
+    pub func: &'a Func,
+    pub store: Option<&'a Store<F>>,
+    pub input: Option<Vec<Ptr<F>>>,
+    pub output: Option<Vec<Ptr<F>>>,
+    pub frames: Option<Vec<Frame<F>>>,
+    pub cached_witness: Option<WitnessCS<F>>,
+    pub count: usize,
+}
```

Comment:
Have you thought of having a trait `MultiFrameTrait<F>`, with associated types `Store`, `Witness`, `Ptr` and `Frame`, and giving it two implementations? 
Then the code downstream of this (notably the nova prover) can call into the API of the *trait*, instead of concrete instantiations.

This seems to be the thing that underpins the code duplication downste

Reply:
Yes, although I'm not so sure it's a simple/small task. I think the generalization would have to go deeper, reaching all the way until `ZExprPtr` and `ZContPtr`.

More differences:
* The LEM `MultiFrame` also holds a `&Func`
* The type signature of `synthesize_frames` is very different, potentially motivating a generalization of the global allocator as well